packages/zero-runtime/src/processors/css.ts

```diff
@@ -0,0 +1,179 @@
+import type { Expression } from '@babel/types';
+import { validateParams } from '@linaria/tags';
+import type {
+  CallParam,
+  TemplateParam,
+  Params,
+  TailProcessorParams,
+  ValueCache,
+} from '@linaria/tags';
+import type { Replacements, Rules } from '@linaria/utils';
+import { ValueType } from '@linaria/utils';
+import type { CSSInterpolation } from '@emotion/css';
+import deepMerge from 'lodash.merge';
+import BaseProcessor from './base-processor';
+import type { IOptions } from './styled';
+import { cache, css } from '../utils/emotion';
+
+type Primitive = string | number | boolean | null | undefined;
+
+/**
+ * @description Scope css class generation similar to css from emotion.
+ *
+ * @example
+ * ```ts
+ * import { css } from '@mui/zero-runtime';
+ *
+ * const class1 = css(({theme}) => ({
+ *  color: (theme.vars || theme).palette.primary.main,
+ * }))
+ * ```
+ *
+ * <html className={class1} />
+ */
+export class CssProcessor extends BaseProcessor {
+  callParam: CallParam | TemplateParam;
+
+  constructor(params: Params, ...args: TailProcessorParams) {
+    super(params, ...args);
+    if (params.length < 2) {
+      throw BaseProcessor.SKIP;
+    }
+    validateParams(
+      params,
+      ['callee', ['call', 'template']],
+      `Invalid use of ${this.tagSource.imported} tag.`,
+    );
+
+    const [, callParams] = params;
+    if (callParams[0] === 'call') {
+      const [, ...callArgs] = callParams;
+      this.dependencies.push(...callArgs);
+    } else if (callParams[0] === 'template') {
+      callParams[1].forEach((element) => {
+        if ('kind' in element && element.kind !== ValueType.CONST) {
+          this.dependencies.push(element);
+        }
+      });
+    }
+    this.callParam = callParams;
+  }
+
+  build(values: ValueCache) {
+    if (this.artifacts.length > 0) {
+      throw new Error(`MUI: "${this.tagSource.imported}" is already built`);
+    }
+
+    const [callType] = this.callParam;
+
+    if (callType === 'template') {
+      this.handleTemplate(this.callParam, values);
+    } else {
+      this.handleCall(this.callParam, values);
+    }
+  }
+
+  private handleTemplate([, callArgs]: TemplateParam, values: ValueCache) {
+    const templateStrs: string[] = [];
+    const templateExpressions: Primitive[] = [];
+    callArgs.forEach((item) => {
+      if ('kind' in item) {
+        switch (item.kind) {
+          case ValueType.FUNCTION:
+            throw item.buildCodeFrameError(
+              'MUI: Functions are not allowed to be interpolated in css tag.',
```

Comment:
If my understanding is correct, functions are not allowed in a literal string? Eg, this will not work?

```js
css`
  display: ${({ theme }) => props.disabled ? theme.vars.palette.text.disabled : theme.vars.palette.text.primary};
`
```

Reply:
Here, where will you be getting the `props` from ? This is the main reason I did not add callback support.
If it is just about the `theme` argument, then I can add this support.

BTW, it is also possible to do something like this -

```js
css(({theme}) => `
  display: ${theme.vars.palette.text.disabled};
`
```

Reply:
> BTW, it is also possible to do something like this -

Okay, I think that is enough.